README
v1.0 2016-8-17
1.实现多进程logs日志实时写入文件。
2.实现程序启动先读取配置文件。
3.实现主进程生成PID文件和主进程log文件，主进程监控logs文件清理。
4.实现根据开始时间和结束时间统计ping值信息。

----------------------------------------
v2.0 2016-8-29
1.实现主进程启动检查现有进程是否已经启动，如重复启动，打印错误日志。
2.实现主进程接收指令"-d"以后台进程方式启动，"-r"重启主进程，"-s"停止主进程。
3.增加子进程检查功能，主进程不存在，子进程自动退出。

----------------------------------------
v2.1 2016-8-30
1.修复timeout会出现2秒延迟。
2.减少连续时间的秒数被中断的次数。

----------------------------------------
v2.2 2016-8-31
1.调整发送ping包的算法，精确每秒发送一个。

----------------------------------------
v2.3 2016-9-13
1.调整每小时删除一次日志的代码，改为删除日志文件的时候才输出info信息。
2.修改"-r"参数，主进程停止1秒后再启动进程。

----------------------------------------
v2.4 2016-11-10
1.修改主进程和子进程的proctitl信息，既使用ps显出来的"COMMAND"，用于区别进程。
2.修改帮助信息，显示程序版本。
3.修改第三方个库的安装方式，直接从libs里复制到/usr/local/python2.6/site-packages/下。

----------------------------------------
v2.5 2017-1-19
1.修复出现ping值出现超大值的现象，设置ping值大于4秒的默认为timeout。
2.更新contrast的帮助信息。
3.更新pings和contrast的版本号。

----------------------------------------
v2.6 2017-2-8
1.调整程序调用第三方库的路径，直接调用程序当前目录./lib目录里面的第三方包。

----------------------------------------
v2.7 2017-2-15
1.修复找不到start_time时的逻辑bug。
2.统一文件编码为utf-8，print函数里面的内容不添加u前缀。
3.增加--help参数的帮助信息入口。

----------------------------------------
v2.8 2017-3-3
1.配置文件增加标识[Host_tag]，字段tag。
2.contrast增加当前主机的标识。
3.调整pings子进程函数（ping_write_log）的前后时间（before、after）逻辑。

----------------------------------------
2017-5-12
pings.py v2.9
1.pings.py增加logging模块输出log。
2.调整pings.py的global_init(),减少全局函数，降低函数间的耦合性。
3.优化pings.py里的主进程和子进程的逻辑。

contrast.py v2.9
1.contrast.py增加logging模块输出log。

view.py v1.0
1.新增view.py提供给web查询对应的log信息。

----------------------------------------
2017-5-16
pings.py v2.10
1.增加子进程异常处理，异常情况：子进程打开的log文件被删除的情况下的异常处理。
2.增加重启参数在屏幕打印的信息。

CHANGELOG.md
# 历史更新记录
    v1.0 2016-8-17
    1.实现多进程logs日志实时写入文件。
    2.实现程序启动先读取配置文件。
    3.实现主进程生成PID文件和主进程log文件，主进程监控logs文件清理。
    4.实现根据开始时间和结束时间统计ping值信息。

    ----------------------------------------
    v2.0 2016-8-29
    1.实现主进程启动检查现有进程是否已经启动，如重复启动，打印错误日志。
    2.实现主进程接收指令"-d"以后台进程方式启动，"-r"重启主进程，"-s"停止主进程。
    3.增加子进程检查功能，主进程不存在，子进程自动退出。

    ----------------------------------------
    v2.1 2016-8-30
    1.修复timeout会出现2秒延迟。
    2.减少连续时间的秒数被中断的次数。

    ----------------------------------------
    v2.2 2016-8-31
    1.调整发送ping包的算法，精确每秒发送一个。

    ----------------------------------------
    v2.3 2016-9-13
    1.调整每小时删除一次日志的代码，改为删除日志文件的时候才输出info信息。
    2.修改"-r"参数，主进程停止1秒后再启动进程。

    ----------------------------------------
    v2.4 2016-11-10
    1.修改主进程和子进程的proctitl信息，既使用ps显出来的"COMMAND"，用于区别进程。
    2.修改帮助信息，显示程序版本。
    3.修改第三方个库的安装方式，直接从libs里复制到/usr/local/python2.6/site-packages/下。

    ----------------------------------------
    v2.5 2017-1-19
    1.修复出现ping值出现超大值的现象，设置ping值大于4秒的默认为timeout。
    2.更新contrast的帮助信息。
    3.更新pings和contrast的版本号。

    ----------------------------------------
    v2.6 2017-2-8
    1.调整程序调用第三方库的路径，直接调用程序当前目录./lib目录里面的第三方包。

    ----------------------------------------
    v2.7 2017-2-15
    1.修复找不到start_time时的逻辑bug。
    2.统一文件编码为utf-8，print函数里面的内容不添加u前缀。
    3.增加--help参数的帮助信息入口。

    ----------------------------------------
    v2.8 2017-3-3
    1.配置文件增加标识[Host_tag]，字段tag。
    2.contrast增加当前主机的标识。
    3.调整pings子进程函数（ping_write_log）的前后时间（before、after）逻辑。

    ----------------------------------------
    2017-5-12
    pings.py v2.9
    1.pings.py增加logging模块输出log。
    2.调整pings.py的global_init(),减少全局函数，降低函数间的耦合性。
    3.优化pings.py里的主进程和子进程的逻辑。

    contrast.py v2.9
    1.contrast.py增加logging模块输出log。

    view.py v1.0
    1.新增view.py提供给web查询对应的log信息。

    ----------------------------------------
    2017-5-16
    pings.py v2.10
    1.增加子进程异常处理，异常情况：子进程打开的log文件被删除的情况下的异常处理。
    2.增加重启参数在屏幕打印的信息。

view.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys
import datetime
import os
import logging

__author__ = 'L'
__version__ = "1.0"


def show_help():
    """
    显示帮助信息。
    :return:
    """
    print("View version v%s" % __version__)
    print("Usage: python view.py --host=test1 --date=20170512 --start_time=00:00:01 --end_time=01:00:00")


def view_log(host, date, start_time, end_time):
    start_flag = False
    end_flag = False
    start = datetime.datetime.strptime(start_time, "%X")
    end = datetime.datetime.strptime(end_time, "%X")

    host_dir = LOGS_PATH + os.sep + host
    if os.path.exists(host_dir):
        log_file = host_dir + os.sep + host + "_" + date + ".log"
        if os.path.exists(log_file):
            if end > start:
                with open(log_file, "r") as log_f:
                    for num, line in enumerate(log_f):
                        line_list = line.split()
                        if len(line_list) == 4:
                            try:
                                time_temp = datetime.datetime.strptime(line_list[1], "%X")
                            except ValueError:
                                logging.error("line:%d, Can not change time '%s' form string to datetime." % (num, line_list[1]))
                            if time_temp == start:
                                start_flag = True
                            if time_temp >= end:
                                end_flag = True
                            if start_flag:
                                print(line.rstrip())
                                if end_flag:
                                    break
                if not start_flag:
                    logging.error("Can not found 'start_time'.")
            else:
                logging.error("'start_time' bigger than 'end_time'.")
        else:
            logging.error("Can not found file '%s'." % log_file)
    else:
        logging.error("Can not found the host '%s'" % host)


def main():
    global LOGS_PATH
    global CONFIG_PATH
    LOGS_PATH = sys.path[0] + os.sep + "logs"
    CONFIG_PATH = sys.path[0] + os.sep + "pings.conf"

    argv_host = ""
    argv_date = ""
    argv_start_time = ""
    argv_end_time = ""

    log_format = "[%(asctime)s] [line:%(lineno)-3d] [%(levelname)-8s] %(message)s"
    logging.basicConfig(level=logging.WARNING, format=log_format, datefmt="%Y-%m-%d %H:%M:%S")
    logging.info("Start main()")

    argument_len = len(sys.argv)
    if argument_len == 5:
        for i in range(1, argument_len):
            str1 = (sys.argv[i]).split("=")[0]
            str2 = (sys.argv[i]).split("=")[1]
            if str1 == "--host":
                argv_host = str2
            if str1 == "--date":
                argv_date = str2
            if str1 == "--start_time":
                argv_start_time = str2
            if str1 == "--end_time":
                argv_end_time = str2
        if argv_host != "" and argv_date != "" and argv_start_time != "" and argv_end_time != "":
            try:
                datetime.datetime.strptime(argv_date, "%Y%m%d").date()
            except ValueError:
                logging.error("Argument 'date' value error.")
                sys.exit(1)
            try:
                datetime.datetime.strptime(argv_start_time, "%X").time()
            except ValueError:
                logging.error("Argument 'start_time' value error.")
                sys.exit(1)
            try:
                datetime.datetime.strptime(argv_end_time, "%X").time()
            except ValueError:
                logging.error("Argument 'end_time' value error.")
                sys.exit(1)
            # argv_host = "test1"
            # argv_date = "20170511"
            # argv_start_time = "18:30:00"
            # argv_end_time = "18:30:15"
            view_log(argv_host, argv_date, argv_start_time, argv_end_time)
    elif argument_len == 2:
        if sys.argv[1] == "--help":
            show_help()
        else:
            logging.error("Need 4 arguments.")
            show_help()
    else:
        logging.error("Need 4 arguments.")
        show_help()


if __name__ == "__main__":
    main()


contrast.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

from __future__ import division
import os
import sys
import datetime
import ConfigParser
import logging

sys.path.append(sys.path[0] + os.sep + "lib")
from prettytable import PrettyTable

__author__ = 'L'
__version__ = "2.9"


def count(date, start_time, end_time):
    """
    统计ping信息并生成表格。
    :param date: 字符串，需要统计的日期，格式：--date=20160816。
    :param start_time: 字符串，统计开始的时间点，格式：--start_time=01:00:00。
    :param end_time:字符串，统计结束的时间点，格式：--end_time=09:00:00。
    :return:
    """
    logging.info("turn to read_config()")
    tag, name_list, ip_list, remark_list = read_config()
    table = PrettyTable(["Ping(%s)" % tag, "Total time(ms)", "Avg time(ms)", "Total icmp", "Loss icmp", "Loss percent(%)"])
    table.padding_width = 1

    for (i, n) in enumerate(name_list):
        total_time = 0.0
        total_count = 0
        loss_count = 0
        start_time_exist = False
        end_time_exist = False
        file_path = LOGS_PATH + os.sep + n + os.sep + n + "_" + date + ".log"
        if os.path.exists(file_path):
            logging.info("open file '%s', start count." % file_path)
            with open(file_path, "r") as log_file:
                for num, line in enumerate(log_file):
                    line_list = line.split(" ")
                    if len(line_list) == 4:
                        if line_list[1] == start_time:
                            start_time_exist = True
                        if line_list[1] == end_time:
                            end_time_exist = True
                        if start_time_exist:
                            total_count += 1
                            delay = line_list[3].rstrip()
                            if delay == "timeout":
                                loss_count += 1
                            else:
                                try:
                                    total_time += float(delay)
                                except ValueError:
                                    logging.error("File:'%s' line:%d Can not change 'delay' value(%s) from str to float." % (file_path, num, delay))
                                    sys.exit(3)
                        if end_time_exist:
                            break
                    else:
                        logging.error("File:'%s' line:%d data incomplete." % (file_path, num))
                        sys.exit(4)
                if start_time_exist and end_time_exist:
                    name = ip_list[i] + "(" + remark_list[i] + ")"
                    avg_time = total_time / total_count
                    loss_percent = loss_count / total_count * 100
                    table.add_row([name, format(total_time, ".1f"), format(avg_time, ".2f"), total_count, loss_count, format(loss_percent, ".2f")])
                elif not start_time_exist:
                    logging.error("HOST:'%s' IP:'%s' REMARK:'%s', start_time:'%s' not found." % (name_list[i], ip_list[i], remark_list[i], start_time))
                else:
                    logging.error("HOST:'%s' IP:'%s' REMARK:'%s', end_time:'%s' not found." % (name_list[i], ip_list[i], remark_list[i], end_time))
        else:
            logging.warning("File '%s' not found." % file_path)
    print(table)


def read_config():
    """
    读取配置文件，获取name_list、ip_list、remark_list。
    :return tag: 字符串类型，配置文件里主机标识信息。
    :return name_list: 字符串列表，配置文件里IP地址列表的名称。
    :return ip_list: 字符串列表，配置文件里IP地址列表。
    :return remark_list: 字符串列表，配置文件里remark列表。
    """
    tag = ""
    name_list = []
    ip_list = []
    remark_list = []
    config = ConfigParser.ConfigParser()

    if os.path.exists(CONFIG_PATH):
        config.read(CONFIG_PATH)
        if len(config.sections()) > 0:
            for section in config.sections():
                if section == "Host_tag":
                    tag = config.get(section, "tag")
                elif section != "Logs":
                    name_list.append(section)
                    ip_list.append(config.get(section, "ip"))
                    remark_list.append(config.get(section, "remark"))
            return [tag, name_list, ip_list, remark_list]
    else:
        logging.error("Can not find the configuration file '%s'." % CONFIG_PATH)
        sys.exit(2)


def show_help():
    """
    显示帮助信息。
    :return:
    """
    print("Contrast version：v%s" % __version__)
    print("Usage: python contrast.py --date=20170511 --start_time=00:00:01 --end_time=09:00:00")


def main():
    """
    主程序入口，检查参数，启动统计函数。
    :return:
    """
    # 定义log级别CRITICAL,ERROR,WARNING,INFO,DEBUG，默认设置为WARNING
    log_format = "[%(asctime)s] [line:%(lineno)-3d] [%(levelname)-8s] %(message)s"
    logging.basicConfig(level=logging.WARNING, format=log_format, datefmt="%Y-%m-%d %H:%M:%S")
    logging.info("Start main()")

    global LOGS_PATH
    global CONFIG_PATH
    argv_date = ""
    argv_start_time = ""
    argv_end_time = ""
    root_path = sys.path[0]
    LOGS_PATH = root_path + os.sep + "logs"
    CONFIG_PATH = root_path + os.sep + "pings.conf"

    logging.info("judge arguments.")
    argument_len = len(sys.argv)
    if argument_len == 4:
        for i in range(1, argument_len):
            str1 = (sys.argv[i]).split("=")[0]
            str2 = (sys.argv[i]).split("=")[1]
            if str1 == "--date":
                argv_date = str2
            if str1 == "--start_time":
                argv_start_time = str2
            if str1 == "--end_time":
                argv_end_time = str2
        if argv_date != "" and argv_start_time != "" and argv_end_time != "":
            try:
                start = datetime.datetime.strptime(argv_start_time, "%X").time()
            except ValueError:
                logging.error("Argument 'start_time' value error.")
                sys.exit(1)
            try:
                end = datetime.datetime.strptime(argv_end_time, "%X").time()
            except ValueError:
                logging.error("Argument 'end_time' value error.")
                sys.exit(1)
            start_seconds = start.hour * 3600 + start.minute * 60 + start.second
            end_seconds = end.hour * 3600 + end.minute * 60 + end.second
            if end_seconds - start_seconds > 0:
                logging.info("Turn to count()")
                count(argv_date, argv_start_time, argv_end_time)
            else:
                logging.error("'start_time' bigger than 'end_time'.")
        else:
            logging.error("'date' or 'start_time' or 'end_time' can not be empty.")
            show_help()
    elif argument_len == 2:
        if sys.argv[1] == "--help":
            show_help()
    elif argument_len == 1:
        show_help()
    else:
        logging.error("Need 3 arguments.")
        show_help()


if __name__ == "__main__":
    main()


pings.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import time
import os
import ConfigParser
import datetime
import multiprocessing
import sys
import platform
import logging

sys.path.append(sys.path[0] + os.sep + "lib")
import ping
import prctl

__author__ = "L"
__version__ = "2.10"


def global_init():
    """
    定义主进程全局常量和logging配置。
    :return:
    """
    global ROOT_PATH
    global LOGS_PATH
    global CONFIG_PATH
    global MAIN_LOG_PATH
    global PID_PATH
    ROOT_PATH = sys.path[0]
    LOGS_PATH = ROOT_PATH + os.sep + "logs"
    CONFIG_PATH = ROOT_PATH + os.sep + "pings.conf"
    MAIN_LOG_PATH = ROOT_PATH + os.sep + "pings.log"
    PID_PATH = ROOT_PATH + os.sep + "pings.pid"

    log_format = "[%(asctime)s] [line:%(lineno)-3d] [%(levelname)-8s] %(message)s"
    logging.basicConfig(level=logging.INFO, format=log_format, datefmt="%Y-%m-%d %H:%M:%S", filename=MAIN_LOG_PATH)


def read_config():
    """
    读取配置文件信息，判断配置是否合法，并返回配置信息。
    :return flag：布尔型，True表示配置文件合法，False表示配置文件不合法。
    :return cycle：整形，logs目录里面各个子目录下的*.log文件保存的周期。
    :return name_list：字符串列表，表示配置文件里面所有IP地址列表的名称。
    :return ip_list: 字符串列表，配置文件里名称对应的IP列表。
    :return remark_list: 字符串列表，配置文件里名称对应的remark列表。
    """
    flag = False
    cycle = 7
    name_list = []
    ip_list = []
    remark_list = []

    config = ConfigParser.ConfigParser()
    if os.path.exists(CONFIG_PATH):
        config.read(CONFIG_PATH)
        if len(config.sections()) > 0:
            for section in config.sections():
                if section != "Host_tag":
                    option_count = len(config.options(section))
                    if section == "Logs" and option_count == 1:
                        cycle = config.get("Logs", "clean_cycle")
                    elif section != "Logs" and option_count == 2:
                        name_list.append(section)
                        ip_list.append(config.get(section, "ip"))
                        remark_list.append(config.get(section, "remark"))
            if len(name_list) == 0 or ip_list == 0 or remark_list == 0:
                return [flag, cycle, name_list, ip_list, remark_list]
            else:
                flag = True
                return [flag, cycle, name_list, ip_list, remark_list]
        else:
            return [flag, cycle, name_list, ip_list, remark_list]
    else:
        return [flag, cycle, name_list, ip_list, remark_list]


def check_logs_path(name_list):
    """
    判断logs文件夹是否存在。
    :param name_list：字符串列表，表示配置文件里面所有IP地址列表的名称。
    :return :flag：布尔型，True表示成功创建logs目录和对应的子目录，False表示失败。
    """
    if os.path.exists(LOGS_PATH):
        for name in name_list:
            logs_name_path = LOGS_PATH + os.sep + name
            if not os.path.exists(logs_name_path):
                os.mkdir(logs_name_path)
        flag = True
    else:
        os.mkdir(LOGS_PATH)
        for name in name_list:
            logs_name_path = LOGS_PATH + os.sep + name
            os.mkdir(logs_name_path)
        flag = True
    return flag


def ping_write_log(name, dst_ip, ppid):
    """
    子进程函数，ping目标IP，并写入log文件，log文件为  名称_日期.log，例如：test1_20160815.log。
    :param name：字符串，配置文件里IP地址列表的名称。
    :param dst_ip：字符串，需要ping的目标IP地址。
    :param ppid: 字符串，父进程的PID。
    :return:
    """
    first_day = datetime.datetime.now().date()
    prctl.set_proctitle("ping " + name)
    while True:
        if check_pid(ppid):
            day_temp = datetime.datetime.now().date()
            log_path = LOGS_PATH + os.sep + name + os.sep + name + "_" + datetime.datetime.now().strftime("%Y%m%d") + ".log"
            with open(log_path, mode="a") as log_file:
                before = datetime.datetime.now()
                while first_day == day_temp:
                    if check_pid(ppid):
                        after = datetime.datetime.now()
                        if (after.hour * 3600 + after.minute * 60 + after.second) > (before.hour * 3600 + before.minute * 60 + before.second):
                            time_str = after.strftime("%Y-%m-%d %X")
                            delay = ping.do_one(dst_ip, timeout=0.7, psize=64)
                            if delay is None or delay > 4.0:
                                if os.path.exists(log_path):
                                    log_file.write(time_str + " " + dst_ip + " timeout\n")
                                    log_file.flush()
                                else:
                                    break
                            else:
                                delay_ms = delay * 1000
                                if os.path.exists(log_path):
                                    log_file.write(time_str + " " + dst_ip + " " + str(format(delay_ms, ".1f")) + "\n")
                                    log_file.flush()
                                else:
                                    break
                            before = datetime.datetime.now()
                        else:
                            before = datetime.datetime.now()
                            time.sleep(0.1)
                        day_temp = datetime.datetime.now().date()
                    else:
                        logging.info("Main process exit.")
                        logging.info("Subprocess exit. PPID:%s, PID:%s, HOST:%s, IP:%s" % (ppid, os.getpid(), name, dst_ip))
                        return
            if day_temp >= first_day:
                first_day = day_temp
            else:
                logging.error("Wrong date. Subprocess exit. PPID:%s, PID:%s, HOST:%s, IP:%s" % (ppid, os.getpid(), name, dst_ip))
                return
        else:
            logging.info("Main process exit.")
            logging.info("Subprocess exit. PPID:%s, PID:%s, HOST:%s, IP:%s" % (ppid, os.getpid(), name, dst_ip))
            return


def create_pid_file(name_list, ppid, process_list):
    """
    生成pings.pid文件，pid文件里面第一行是主进程的PID，后面是子进程对应的PID，匹配配置文件里的IP名称列表。
    :param name_list: 字符串列表，主机名列表。
    :param ppid: 字符串列表，主进程的PID。
    :param process_list: 字符串列表，启动的子进程对象。
    :return:
    """
    logging.info("Write pid file.")
    with open(PID_PATH, mode="w") as pid_file:
        pid_file.write(ppid + " PPID" + "\n")
        for i in range(0, len(process_list)):
            pid_str = str(process_list[i].pid) + " " + name_list[i] + "\n"
            pid_file.write(pid_str)


def clean_log(cycle, name_list):
    """
    定期清除log文件
    :param cycle: 整型，保存多少天的log。
    :param name_list: 字符串列表，主机名称列表。
    :return:
    """
    date_file = ""
    clean_cycle = datetime.timedelta(days=int(cycle))
    date_now = datetime.datetime.now().date()
    for n in name_list:
        log = LOGS_PATH + os.sep + n
        for f in os.listdir(log):
            if f[-3:] == "log":
                log_date = f[-12:-4]
                try:
                    date_file = datetime.datetime.strptime(log_date, "%Y%m%d").date()
                except ValueError:
                    logging.warning("Can not change the logs file name('%s') form str to datetime." % f)
                if date_now - date_file >= clean_cycle:
                    os.remove(log + os.sep + f)


def background():
    """
    后台启动主进程函数。
    :return:
    """
    logging.debug("Turn to check_process().")
    if not check_process():
        logging.info("Main process start.")
        logging.debug("Turn to read_config().")
        config_flag, cycle, name_list, ip_list, remark_list = read_config()
        logging.info("Check config.")
        if config_flag:
            print("[INFO] Program start.")
            logging.info("Check config OK.")
            logging.debug("Turn to daemonize().")
            # daemonize('/dev/null', '/tmp/daemon_stdout.log', '/tmp/daemon_error.log')
            daemonize()
            logging.debug("Turn to start_process().")
            start_process(cycle, name_list, ip_list)
        else:
            logging.error("Config file '%s' read error. Main process stop." % CONFIG_PATH)
            print("[ERROR] Config file '%s' read error. Program exit." % CONFIG_PATH)
    else:
        print("[ERROR] Program is running. Please do not run the program repeatedly.")


def start_process(cycle, name_list, ip_list):
    """
    启动子进程，执行主进程循环，清理过期日志文件。
    :param cycle: 整型，保存多少天的log。
    :param name_list: 字符串列表，主机名称列表。
    :param ip_list: 字符串列表，IP地址列表。
    :return:
    """
    process_list = []

    logging.info("Check logs directory structure.")
    logging.debug("Turn to check_logs_path().")
    if check_logs_path(name_list):
        logging.info("Check logs directory structure OK.")
        logging.info("Begin to start subprocess.")
        ppid = str(os.getpid())
        for (i, n) in enumerate(name_list):
            process = multiprocessing.Process(name=n, target=ping_write_log, args=(n, ip_list[i], ppid))
            process.daemon = True
            process.start()
            logging.info("Start subprocess, PPID:%s, PID:%s, HOST:%s, IP:%s." % (ppid, str(process.pid), n, ip_list[i]))
            process_list.append(process)

        logging.debug("Turn to create_pid_file().")
        create_pid_file(name_list, ppid, process_list)

        first_day = datetime.datetime.now().date()
        prctl.set_proctitle(prctl.get_name() + " " + sys.argv[0] + " -d")
        logging.info("Main process run loop task.")
        while True:
            day_temp = datetime.datetime.now().date()
            if day_temp > first_day:
                first_day = day_temp
                logging.info("Clean expired logs.")
                clean_log(cycle, name_list)
                logging.info("Clean expired logs OK.")
            else:
                time.sleep(3600)
    else:
        logging.error("Check logs directory structure error. Program exit.")


def stop_process():
    """
    停止主进程。
    :return:
    """
    logging.info("Stopping main process.")
    logging.info("Check PID file.")
    if os.path.exists(PID_PATH):
        with open(PID_PATH, mode="r") as pid_file:
            line = pid_file.readline()
            l = line[:-1].split(" ")
            logging.debug("Turn to check_pid().")
            if check_pid(l[0]):
                print("[INFO] Program stop.")
                os.kill(int(l[0]), 15)
                logging.info("Main process stopped.")
            else:
                logging.info("Main process not found. Program exit.")
    else:
        logging.info("Pid file 'pings.pid' not found. Program exit.")


def check_pid(pid):
    """
    检查进程PID，判断进程是否存在。
    :param pid: 字符串，进程的PID。
    :return: 布尔值，True为进程存在，False为进程不存在。
    """
    try:
        os.kill(int(pid), 0)
    except OSError:
        return False
    else:
        return True


def check_process():
    """
    检查主进程是否存在。
    :return: 布尔值
    """
    logging.info("Check main process.")
    if os.path.exists(PID_PATH):
        with open(PID_PATH, mode="r") as pid_file:
            line = pid_file.readline()
            l = line[:-1].split(" ")
            return check_pid(l[0])
    else:
        return False


def daemonize(stdin='/dev/null', stdout='/dev/null', stderr='/dev/null'):
    """
    守护进程函数。
    :param stdin: 字符串，重定向标准文件描述符。
    :param stdout: 字符串，重定向标准文件描述符。
    :param stderr: 字符串，重定向标准文件描述符。
    :return:
    """
    try:
        pid = os.fork()
        if pid > 0:
            sys.exit(0)
    except OSError:
        sys.exit(1)

    os.chdir("/")
    os.umask(022)
    os.setsid()

    try:
        pid = os.fork()
        if pid > 0:
            sys.exit(0)
    except OSError:
        sys.exit(1)

    for f in sys.stdout, sys.stderr:
        f.flush()
    si = open(stdin, 'r')
    so = open(stdout, 'a+')
    se = open(stderr, 'a+', 0)
    os.dup2(si.fileno(), sys.stdin.fileno())
    os.dup2(so.fileno(), sys.stdout.fileno())
    os.dup2(se.fileno(), sys.stderr.fileno())


def usage():
    """
    打印使用方法。
    :return:
    """
    print("Pings version：v%s" % __version__)
    print("Usage: python pings.py [-d|-s|-r]")
    print("Options:")
    print("     -d: run in daemon mode.")
    print("     -s: stop the program.")
    print("     -r: restart the program.")


def main():
    """
    程序入口。
    :return:
    """
    global_init()

    if platform.system() == "Linux":
        if len(sys.argv) == 1:
            usage()
        elif len(sys.argv) == 2:
            if sys.argv[1] == "-s":
                logging.debug("Turn to stop_process().")
                stop_process()
            elif sys.argv[1] == "-r":
                logging.debug("Turn to stop_process().")
                stop_process()
                time.sleep(1)
                logging.debug("Turn to background().")
                background()
            elif sys.argv[1] == "-d":
                logging.debug("Turn to background().")
                background()
            elif sys.argv[1] == "--help":
                usage()
            else:
                print("[ERROR] Wrong argument.")
                usage()
        else:
            print("[ERROR] Wrong argument.")
            usage()
    else:
        print("[ERROR] This program only for linux.")

if __name__ == "__main__":
    main()
